pixelsecurity_sale/models/sale.py

```diff
@@ -39,18 +51,63 @@ def create(self, vals_list):
+            #If the labor line exist, add the hours, else, create the labor line with the hours.
+            if(labor_line):
+                labor_line.product_uom_qty += labor_hours
+            else:
+                self.env['sale.order.line'].create({
+                        'order_id': line.order_id.id,
+                        'product_id': labor_product.id,
+                        'name': labor_product.name,
+                        'product_uom': labor_product.uom_id.id,
+                        'product_template_id': line.product_template_id.labor_product.id,
+                        'product_uom_qty': labor_hours,
+                        'sequence': line.sequence + 1,
+                        'parent_line_id': line.id,
+                        'is_labor_line': True,
+                    })
             new_lines.append(line)
-        return new_lines
\ No newline at end of file
+        #If there is an order, sort the lines.
+        if(order != None):
+            order.sort_lines()
+        return new_lines
+
+    def unlink(self):
+        if self._check_line_unlink():
+            raise UserError(_('You can not remove an order line once the sales order is confirmed.\nYou should rather set the quantity to 0.'))
+        for line in self:
+            #If a labor line exists in the product of the line that will be deleted, substract the corresponding hours from the labor line.
+            if(line.product_template_id.include_labor):
+                labor_product = line.product_template_id.labor_product.product_variant_id
+                labor_line = self.env['sale.order.line'].search([('order_id','=',line.order_id.id),('product_id','=',labor_product.id)])
+                labor_hours = line.product_id.product_tmpl_id.labor_hours * line.product_uom_qty
+                if(labor_line):
+                    #If the substraction of the hours is bigger than 0, then substract them, else, remove the labor line.
+                    if(labor_line.product_uom_qty - labor_hours >= 0):
+                        labor_line.product_uom_qty -= labor_hours
+                    else:
+                        if(labor_line not in self):
+                            labor_line.unlink()
+        return super(SaleOrderLine, self).unlink()
+
+    def write(self, values):
+        #If there's a change in the quantity of the OL, add/substract hours to the corresponding labor line.
+        if('product_uom_qty' in values):
+            if(self.product_id.product_tmpl_id.include_labor):
+                labor_product = self.product_id.product_tmpl_id.labor_product.product_variant_id
+                labor_line = self.env['sale.order.line'].search([('order_id','=',self.order_id._origin.id),('product_id','=',labor_product.id)])
+                delta = values['product_uom_qty'] - self.product_uom_qty 
+                labor_hours = self.product_id.product_tmpl_id.labor_hours * delta
+                labor_line.product_uom_qty += labor_hours
```

Comment:
posiblity of recursion. 
Instead of directly changing product_uom_quantity update values['product_uom_qty'].

Reply:
But the thing is, that labor_line is a different line than the one being changed on values['product_uom_qty'] (self), so if I change the values, it will take effect on a different line than the one intended, I saw recursion happen, that's why I check before if the current line includes_labor, a labor_line would have this value set to false, so It would not call itself again. Is this ok? or should I check another solution to this part of the code? Thanks :)

Reply:
Okay, that makes sense. This shold be good, no need to change it.